### PR TITLE
Add new vehicle sync & functions to the scripting API

### PR DESCRIPTION
**Pull request by Deewarz**
**⚠️ Requires: https://github.com/MafiaHub/Framework/pull/92**

---

## Reverse
- Fix `C_Car::SetActualFuel`
- Rename `C_Car::PosefujZimuVShopu` to `C_Car::SetVehicleDirty` (because its the real name)
- Rename `C_Vehicle::AreBeaconLightsOn` to `C_Vehicle::GetBeaconLightsOn` (consistency and closer to game nomenclature)
- Rename  `C_Vehicle::GetWheelTintColor` to `C_Vehicle::GetWheelColor` (closer to the game nomenclature)
-  `C_Vehicle::GetWheelColor` returns tire color also (field is `m_TireColor`)
-  `C_Vehicle::SetWheelColor` set tire color also 
- Add `C_Vehicle::GetFuelTankCapacity();`
- ~~`C_Vehicle` fields `m_Color1` and `m_Color2` are now `m_MaterialColor` and `m_MaterialColor1` (because its their real names)~~
- `C_Vehicle` `m_fDirtLevel` field  is now `m_Dirty`  (because its the real name)
- `C_Vehicle` `m_fRustLevel` field  is now `m_Rust`  (because its the real name)
- Fix pattern for `C_Vehicle::SetSteer` (use exact)
- Fix pattern for `C_Vehicle::SetWheelColor` (use exact)
- Fix `C_Vector4` fields order

## Vehicles hooks
- Use `hook::get_pattern()` instead of `hook::pattern().get_first()`

## Vehicle debug
- Clean up
- Add rust
- Add license plate
- Fix wheel color (supports for rim and tire)
- Max fuel is now based on `C_Vehicle::GetFuelTankCapacity();`

## Sync
- Use alphabetical order for sync data
- Rename `locked` to `lockState`
- Rename boolean data with suffix `state` by `on` (eg. `hornState` => `hornOn`) (consistency and closer to game nomenclature)
- Rename `radioId` to `radioStationId `
- Always sync `rot` before `pos` (consitency)
- Sync primary color
- Sync secondary color
- Sync dirt
- Sync fuel
- Sync rim color
- Sync rust
- Sync tire color
- Sync window tint

## SDK
- Use alphabetical order
- **BREAKING**: Rename boolean methods suffix `Enabled` by `On` (eg. `getBeaconLightsEnabled` => `getBeaconLightsOn`) (consistency, shorter and closer to game nomenclature)
- **BREAKING**: Rename `getRadioStation` to `getRadioStationId ` and `setRadioStation` to `setRadioStationId `
- Add `getColorPrimary` and `setColorPrimary`
- Add `getColorSecondary` and `setColorSecondary`
- Add `getDirt` and `setDirt`
- Add `getFuel` and `setFuel`
- Add `getRimColor` and `setRimColor`
- Add `getRust` and `setRust`
- Add `getTireColor` and `setTireColor`
- Add `getWindowTint` and `setWindowTint`

## Known issues
- Existing `C_Vehicle::IsEngineOn`, `C_Vehicle::GetSiren`, `C_Vehicle::GetBeaconLightsOn` are not reversed yet. => https://github.com/MafiaHub/MafiaMP/issues/73
- Since we use defaults values when spawning, all vehicles are ~~black~~ white, fuel full, not dirty and not rusty at spawn...

### CODE REVIEW

**Comment by zpl-zak** [comment]
Amazing work done!

as for the `Since we use defaults values when spawning, all vehicles are black, fuel full, not dirty and not rusty at spawn`:
it's a non-issue, since modders can always customize those values at spawn themselves in GM. Our MP should not implement any extra logic to supplement that or to simulate game's spawn behavior.

**Review by zpl-zak — APPROVED** [review]

File: code/client/src/core/hooks/vehicle.cpp [hunk 1]
```diff
@@ -129,22 +130,23 @@ void C_Human2CarWrapper__EndDrive(SDK::C_Human2CarWrapper *pThis, SDK::C_Actor *
...
     MH_CreateHook((LPVOID)C_CarActionBreakIn__TestActionBreakIn_Addr, (PBYTE)C_CarActionBreakIn__TestActionInternal, reinterpret_cast<void **>(&C_CarActionBreakIn__TestActionInternal_original));
 
-    const auto C_CarActionLeave__TestAction_Addr = hook::pattern("40 53 48 83 EC 20 48 8B DA E8 ? ? ? ? 48 8B C8 4C 8B 00 41 FF 90 ? ? ? ? 48 3B D8 75 29 E8 ? ? ? ? 48 8B C8 48 8B 10 FF 92 ? ? ? ? 48 8B C8 BA ? ? ? ? E8 ? ? ? ? 84 C0 0F 94 C0 48 83 C4 20 5B C3 B0 01 48 83 C4 20 5B C3 CC CC CC CC CC CC CC CC CC CC CC CC CC CC CC 41 B0 01 E9 ? ? ? ? CC CC CC CC CC CC CC CC 48 89 5C 24 ? 48 89 6C 24 ?").get_first();
+    const auto C_CarActionLeave__TestAction_Addr = hook::get_pattern("40 53 48 83 EC 20 48 8B DA E8 ? ? ? ? 48 8B C8 4C 8B 00 41 FF 90 ? ? ? ? 48 3B D8 75 29 E8 ? ? ? ? 48 8B C8 48 8B 10 FF 92 ? ? ? ? 48 8B C8 BA ? ? ? ? E8 ? ? ? ? 84 C0 0F 94 C0 48 83 C4 20 5B C3 B0 01 48 83 "
+                                                                     "C4 20 5B C3 CC CC CC CC CC CC CC CC CC CC CC CC CC CC CC 41 B0 01 E9 ? ? ? ? CC CC CC CC CC CC CC CC 48 89 5C 24 ? 48 89 6C 24 ?");
```
Comment (Segfaultd) [inline code/client/src/core/hooks/vehicle.cpp:147]: This shouldn't go on multiple lines with string interpolation

File: code/client/src/core/modules/vehicle.cpp [hunk 2]
```diff
@@ -37,27 +37,46 @@ namespace MafiaMP::Core::Modules {
...
+                    metadata.power           = vehicle->GetPower();
+                    metadata.radioOn         = vehicle->IsRadioOn();
+                    metadata.radioStationId  = vehicle->GetRadioStation();
+                    metadata.rimColor        = {rimColor.r, rimColor.g, rimColor.b, rimColor.a};
+                    metadata.rust            = vehicle->GetVehicleRust();
+                    metadata.sirenOn         = vehicle->GetSiren();
+                    metadata.steer           = vehicle->GetSteer();
+                    metadata.tireColor       = {tireColor.r, tireColor.g, tireColor.b, tireColor.a};
+                    metadata.velocity        = {vehicleVelocity.x, vehicleVelocity.y, vehicleVelocity.z};
+                    metadata.windowTint      = {windowTint.r, windowTint.g, windowTint.b, windowTint.a};
```
Comment (Segfaultd) [inline code/client/src/core/modules/vehicle.cpp:79]: @zpl-zak ok with this? Makes a lot of data to send in the frequent update payload

File: code/client/src/sdk/entities/c_car.cpp [hunk 3]
```diff
@@ -83,6 +75,20 @@ namespace SDK {
...
     }
 
+    void C_Car::SetActualFuel(float fuel) {
+        hook::this_call<void>(gPatterns.C_Car__SetActualFuel, this, fuel);
+
+        // TODO: fix me, doesn't work yet
+        // uint64_t C_Motor = *(uint64_t *)(*(uint64_t *)((uint64_t)this + C_Motor_Offset) + 0x8);
+        // if (C_Motor) {
+        //     hook::this_call(gPatterns.C_Motor__SetFuel, C_Motor, fuel);
+        // }
```
Comment (Segfaultd) [inline code/client/src/sdk/entities/c_car.cpp:85]: Either fix or don't commit temporary comments

File: code/client/src/sdk/entities/c_car.cpp [hunk 4]
```diff
@@ -91,8 +97,8 @@ namespace SDK {
         hook::this_call<void>(gPatterns.C_Car__SetSpeed, this, speed);
     }
 
-    void C_Car::PosefujZimuVShopu(float dirt) {
-        hook::this_call<void>(gPatterns.C_Car__PosefujZimuVShopu, this, dirt);
+    void C_Car::SetVehicleDirty(float dirt) {
+        hook::this_call<void>(gPatterns.C_Car__SetVehicleDirty, this, dirt);
```
Comment (Segfaultd) [inline code/client/src/sdk/entities/c_car.cpp:101]: This shouldn't be renamed. The name is from the game itself, and as "SDK" logic, we have to stick to it. You can declare a custom `SetVehicleDirty` method that calls the `PosefujZimuVShopu` as an alias.

File: code/client/src/sdk/entities/c_car.h [hunk 5]
```diff
@@ -35,17 +32,18 @@ namespace SDK {
...
+        float GetFuelTankCapacity() const;
 
         void SetTransparency(float transparency);
+
         void SetSpeed(float speed);
 
-        /*
-         * PosefujZimuVShopu - same function as SetVehicleDirt
-         */
-        void PosefujZimuVShopu(float dirt);
```
Comment (Segfaultd) [inline code/client/src/sdk/entities/c_car.h:45]: Same here

File: code/client/src/sdk/entities/c_vehicle.cpp [hunk 6]
```diff
@@ -67,21 +67,26 @@ namespace SDK {
...
+    void C_Vehicle::SetVehicleColor(ue::sys::math::C_Vector4 const *color1, ue::sys::math::C_Vector4 const *color2, bool metallic) {
+        hook::this_call(gPatterns.C_Vehicle__SetVehicleColor, this, color1, color2, metallic);
     }
 
     void C_Vehicle::GetVehicleColor(ue::sys::math::C_Vector4 *color1, ue::sys::math::C_Vector4 *color2) const {
-        *color1 = m_Color1;
-        *color2 = m_Color2;
+        *color1 = m_MaterialColor;
+        *color2 = m_MaterialColor1;
     }
```
Comment (Segfaultd) [inline code/client/src/sdk/entities/c_vehicle.cpp:77]: Naming is wrong, please keep 1 and 2

File: code/client/src/sdk/entities/c_vehicle.h [hunk 7]
```diff
@@ -122,31 +122,32 @@ namespace SDK {
...
+        ue::sys::math::C_Vector4 m_MaterialColor1;                         // 0x0BC0 - 0x0BD0 // 3008 - 3024 // 16
+        ue::sys::math::C_Vector4 m_WindowTint;                             // 0x0BD0 - 0x0BE0 // 3024 - 3040 // 16
+        ue::sys::math::C_Vector4 m_WheelColor;                             // 0x0BE0 - 0x0BF0 // 3040 - 3056 // 16
+        ue::sys::math::C_Vector4 m_TireColor;                              // 0x0BF0 - 0x0C00 // 3056 - 3072 // 16
+        char pad_0C08[8];                                                  // 0x0C00 - 0x0C08 // 3072 - 3080 // 8
+        ue::sys::math::C_Vector4 m_InteriorColors[5];                      // 0x0C08 - 0x0C58 // 3080 - 3160 // 80 (5*16)
+        char pad_0C60[8];                                                  // 0x0C58 - 0x0C60 // 3160 - 3168 // 8
+        float m_Dirty;                                                     // 0x0C60 - 0x0C64 // 3168 - 3172 // 4
+        float m_Rust;                                                      // 0x0C64 - 0x0C68 // 3172 - 3176 // 4
+        char pad_14F8[2192];                                               // 0x0C68 - 0x14F8 // 3176 - 5368 // 2192
```
Comment (Segfaultd) [inline code/client/src/sdk/entities/c_vehicle.h:150]: Comments format are wrong

File: code/client/src/sdk/patterns.cpp [hunk 8]
```diff
@@ -15,16 +15,17 @@ namespace SDK {
...
-        gPatterns.C_Car__GetMotorDamage  = hook::get_opcode_address("E8 ? ? ? ? 49 8B CE 0F 28 F8 E8 ? ? ? ? 0F 28 F0");
-        gPatterns.C_Car__Lock            = hook::get_opcode_address("E8 ? ? ? ? 33 C0 48 83 C4 ? 5B C3 ? ? ? ? ? ? ? ? ? ? ? ? ? ? ? B8 ? ? ? ? C3");
-        gPatterns.C_Car__LockEntryPoints = hook::get_opcode_address("E8 ? ? ? ? 45 8D 7E ? 48 8B 4F ?");
+        gPatterns.C_Car__CloseHood           = reinterpret_cast<uint64_t>(hook::get_pattern("48 89 5C 24 ? 48 89 74 24 ? 48 89 7C 24 ? 41 56 48 83 EC ? 48 8B 81 ? ? ? ? 33 DB 48 2B 81 ? ? ? ? 33 FF 48 C1 F8 ? 48 8B F1 4C 63 F0 85 C0 0F 8E ? ? ? "
+                                                                                                      "? 48 89 6C 24 ? 0F 29 74 24 ? 0F 57 F6 66 66 0F 1F 84 00 ? ? ? ? 48 8B 86 ? ? ? ? 48 63 0C B8 48 8B 86 ? ? ? ? 48 8B 14 C8 33 C9 8B 42 ? 39 05 ? ? ? ? "
+                                                                                                      "48 0F 45 CA 48 85 C9 74 ? 48 0F BF 49 ? 48 8B 86 ? ? ? ? 48 8B 14 C8 0F 2F 72 ? 73 ? 0F B6 86 ? ? ? ? 48 8D 8E ? ? ? ? C0 E8 ? 45 33 C0"));
```
Comment (Segfaultd) [inline code/client/src/sdk/patterns.cpp:20]: No string interpolation for patterns

File: code/client/src/sdk/patterns.cpp [hunk 9]
```diff
@@ -399,12 +399,12 @@ namespace SDK {
         gPatterns.C_Vehicle__SetSpeed           = hook::get_opcode_address("E8 ? ? ? ? 49 8B CF E8 ? ? ? ? 45 0F 57 C9");
         gPatterns.C_Vehicle__SetSpeedLimit      = hook::get_opcode_address("E8 ? ? ? ? 48 8B 8B ? ? ? ? 48 85 C9 74 ? 48 8B 89 ? ? ? ?");
         gPatterns.C_Vehicle__SetSPZText         = hook::get_opcode_address("E8 ? ? ? ? 48 8D 5D ? BF ? ? ? ? 0F B7 13");
-        gPatterns.C_Vehicle__SetSteer           = hook::get_opcode_address("E8 ? ? ? ? 45 33 C9 44 89 A7 ? ? ? ?");
+        gPatterns.C_Vehicle__SetSteer           = hook::get_opcode_address("E8 ? ? ? ? 41 B1 01 89 BE ? ? ? ?");
```
Comment (Segfaultd) [inline code/client/src/sdk/patterns.cpp:402]: Were the patterns wrong?

File: code/client/src/sdk/ue/sys/math/c_vector.h [hunk 10]
```diff
@@ -11,10 +11,10 @@ namespace SDK {
 
         class C_Vector4 {
           public:
-            float a;
             float r;
             float g;
             float b;
+            float a;
```
Comment (Segfaultd) [inline code/client/src/sdk/ue/sys/math/c_vector.h:17]: Not sure - 100% sure it's working?

File: code/shared/game_rpc/vehicle/vehicle_setprops.h [hunk 11]
```diff
@@ -7,36 +7,64 @@
...
-            if (radioId.HasValue() && (radioId() >= -1 && radioId() < 4 /* E_RADIO_POLICE */))
+            if (licensePlate.HasValue() && licensePlate().GetLength() > 7) {
                 return false;
-            if (licensePlate.HasValue() && licensePlate().GetLength() > 7)
+            }
+
+            if (radioStationId.HasValue() && (radioStationId() >= -1 && radioStationId() < 4 /* E_RADIO_POLICE */)) {
                 return false;
+            }
+
```
Comment (Segfaultd) [inline code/shared/game_rpc/vehicle/vehicle_setprops.h:67]: Please explain

**Review by Segfaultd — CHANGES_REQUESTED** [review]

File: code/client/src/core/modules/vehicle.cpp (hunk 2 above)
Comment (zpl-zak) [reply code/client/src/core/modules/vehicle.cpp:79]: it's fine for now, we will turn these into delta patches later on eventually (bs.SerializeDelta)

File: code/client/src/sdk/entities/c_car.cpp (hunk 3 above)
Comment (Deewarz) [reply code/client/src/sdk/entities/c_car.cpp:85]: This was the original code that didn't work, ok to remove it?

File: code/client/src/sdk/entities/c_car.cpp (hunk 4 above)
Comment (Deewarz) [reply code/client/src/sdk/entities/c_car.cpp:101]: image

 From M3, same code, pretty sure PosefujZimuVShopu has been removed in favor of SetVehicleDirty

File: code/client/src/core/hooks/vehicle.cpp (hunk 1 above)
Comment (Deewarz) [reply code/client/src/core/hooks/vehicle.cpp:147]: Can't, this modification is made by the code formatter

File: code/client/src/sdk/ue/sys/math/c_vector.h (hunk 10 above)
Comment (Deewarz) [reply code/client/src/sdk/ue/sys/math/c_vector.h:17]: Yes

File: code/client/src/core/hooks/vehicle.cpp (hunk 1 above)
Comment (zpl-zak) [reply code/client/src/core/hooks/vehicle.cpp:147]: C/C++ will concat to single line anyways, this should we fine, just ugly.

File: code/shared/game_rpc/vehicle/vehicle_setprops.h (hunk 11 above)
Comment (Deewarz) [reply code/shared/game_rpc/vehicle/vehicle_setprops.h:67]: This code was already there

File: code/client/src/core/hooks/vehicle.cpp (hunk 1 above)
Comment (Segfaultd) [reply code/client/src/core/hooks/vehicle.cpp:147]: Ugly is the point :D Lets fix formatter!

File: code/shared/game_rpc/vehicle/vehicle_setprops.h (hunk 11 above)
Comment (Segfaultd) [reply code/shared/game_rpc/vehicle/vehicle_setprops.h:67]: Didn't you swap the conditions positions?

File: code/client/src/sdk/entities/c_car.h (hunk 5 above)
Comment (Deewarz) [reply code/client/src/sdk/entities/c_car.h:45]: See https://github.com/MafiaHub/MafiaMP/pull/72#discussion_r1469581960

File: code/client/src/sdk/patterns.cpp (hunk 8 above)
Comment (Deewarz) [reply code/client/src/sdk/patterns.cpp:20]: See https://github.com/MafiaHub/MafiaMP/pull/72#discussion_r1469583091

File: code/client/src/sdk/entities/c_car.h (hunk 5 above)
Comment (Segfaultd) [reply code/client/src/sdk/entities/c_car.h:45]: Same answer :-) https://github.com/MafiaHub/MafiaMP/pull/72#discussion_r1469587364

File: code/shared/game_rpc/vehicle/vehicle_setprops.h (hunk 11 above)
Comment (Deewarz) [reply code/shared/game_rpc/vehicle/vehicle_setprops.h:67]: Ah yes (and rename radioId to radioStationId)

Alphabetical order, no change in the logic

File: code/client/src/sdk/patterns.cpp (hunk 9 above)
Comment (Deewarz) [reply code/client/src/sdk/patterns.cpp:402]: Patterns was the function calling in another function (sub_142EC1430).
I used the real pattern of the function : C_Vehicle::SetSteer 

This is the case for others like SetAngularSpeed ​​for example. I didn't change it everywhere.

File: code/client/src/core/hooks/vehicle.cpp (hunk 1 above)
Comment (Deewarz) [reply code/client/src/core/hooks/vehicle.cpp:147]: I'm supposed to do it? 
I admit that I am not expert enough to control the impacts...

File: code/client/src/sdk/entities/c_car.cpp (hunk 3 above)
Comment (Segfaultd) [reply code/client/src/sdk/entities/c_car.cpp:85]: Yes

File: code/client/src/sdk/entities/c_vehicle.cpp [hunk 12]
```diff
@@ -31,8 +31,8 @@ namespace SDK {
         hook::this_call(gPatterns.C_Vehicle__SetHandbrake, this, force, arg2);
     }
 
-    void C_Vehicle::SetEngineOn(bool arg1, bool arg2) {
-        hook::this_call(gPatterns.C_Vehicle__SetEngineOn, this, arg1, arg2);
+    void C_Vehicle::SetEngineOn(bool on, bool arg2) {
+        hook::this_call(gPatterns.C_Vehicle__SetEngineOn, this, on, arg2);
```
Comment (Segfaultd) [inline code/client/src/sdk/entities/c_vehicle.cpp:35]: We don't exactly know what are the two arguments for, sometimes the game pass only the first one and sometimes both. Lets not make assumptions.

**Review by Segfaultd — CHANGES_REQUESTED** [review]

**Pull request merged**